interfaces/inc/c_pal/malloc_multi_flex.h

```diff
@@ -0,0 +1,133 @@
+        pointer_iterator = (uintptr_t)pointer_iterator + MU_C2(arg2, _count) * sizeof(arg1);
+
+    #define MALLOC_MULTI_FLEX_ASSIGN_INTERNAL_STRUCT_PTR_ARRAY_FIELDS(...) \
+        MU_FOR_EACH_2(MALLOC_MULTI_FLEX_ASSIGN_INTERNAL_STRUCT_PTR, __VA_ARGS__)
+
+    #define MALLOC_MULTI_FLEX_ASSIGN_INTERNAL_STRUCT_PTRS(array_fields) \
+            MU_C2A(MALLOC_MULTI_FLEX_ASSIGN_INTERNAL_STRUCT_PTR_, array_fields)
+
+    /* Codes_MALLOC_MULTI_FLEX_24_005: [ MALLOC_MULTI_FLEX shall expand type to the name of the malloc function in the format of: MALLOC_MULTI_FLEX_type. ]*/
+    #define MALLOC_MULTI_FLEX(type) \
+        MU_C2(malloc_multi_flex_, type) \
+
+    #define MALLOC_MULTI_FLEX_DEFINE_FIELD_MEMBER(arg1, arg2) \
+            arg1 arg2;
+
+    #define MALLOC_MULTI_FLEX_DEFINE_MEMBERS_FIELDS(...) \
+            MU_FOR_EACH_2(MALLOC_MULTI_FLEX_DEFINE_FIELD_MEMBER, __VA_ARGS__)
+
+    #define MALLOC_MULTI_FLEX_DEFINE_ARRAY_FIELD_MEMBER(arg1, arg2) \
+                arg1* arg2;
+
+    #define MALLOC_MULTI_FLEX_DEFINE_MEMBERS_ARRAY_FIELDS(...) \
+                MU_FOR_EACH_2(MALLOC_MULTI_FLEX_DEFINE_ARRAY_FIELD_MEMBER, __VA_ARGS__)
+
+    #define GENERATE_MULTI_MALLOC_STRUCT(fields) \
+            MU_C2A(MALLOC_MULTI_FLEX_DEFINE_MEMBERS_, fields) \
+
+    #define DECLARE_MALLOC_MULTI_FLEX(type, fields, array_fields)\
+        typedef struct MU_C2(type, _TAG) \
+        { \
+            GENERATE_MULTI_MALLOC_STRUCT(fields)\
+            GENERATE_MULTI_MALLOC_STRUCT(array_fields)\
+        } type; \
+        MOCKABLE_FUNCTION(, void*, MALLOC_MULTI_FLEX(type), size_t, parent_struct_size MALLOC_MULTI_FLEX_DECLARE_ARG_LIST_VALUES(array_fields)); \
+
+    #define DEFINE_MALLOC_MULTI_FLEX(type, fields, array_fields)\
+        void* MALLOC_MULTI_FLEX(type)(size_t parent_struct_size MALLOC_MULTI_FLEX_DEFINE_ARG_LIST_VALUES(array_fields)) \
+        {\
+            size_t size_required = parent_struct_size;\
+            /* Codes_SRS_MALLOC_MULTI_FLEX_24_001: [ If the total amount of memory required to allocate the type along with its members exceeds SIZE_MAX then DEFINE_MALLOC_MULTI_FLEX shall fail and return NULL. ]*/ \
+            MALLOC_MULTI_FLEX_ARGS_OVERFLOW_CHECK(array_fields)\
+            /* Codes_SRS_MALLOC_MULTI_FLEX_24_002: [ DEFINE_MALLOC_MULTI_FLEX shall call malloc to allocate memory for the struct and its members. ]*/ \
+            type* parent_struct_pointer = malloc(size_required);\
+            if (parent_struct_pointer == NULL)\
+            {\
+                LogError("malloc(size_required = %zu) failed", size_required);\
+                return NULL;\
+            }\
+            uintptr_t pointer_iterator = (uintptr_t)parent_struct_pointer + parent_struct_size; \
+            MALLOC_MULTI_FLEX_ARGS_VALIDATION(array_fields)\
+            /* Codes_SRS_MALLOC_MULTI_FLEX_24_003: [ DEFINE_MALLOC_MULTI_FLEX shall assign address pointers to all the member arrays. ]*/ \
```

Comment:
I think this leaks. Should have been caught by unit tests.

Reply:
It seems like all checks should be done before we malloc.

Reply:
>I think this leaks. Should have been caught by unit tests.

As per the code 
```c
#define MALLOC_MULTI_FLEX_ARG_VALIDATION(arg1, arg2) \
        if (MU_C2(arg2, _count) > 0 && sizeof(arg1) != sizeof(parent_struct_pointer->arg2[0]))\
        {\
            LogError("Data type size type mistmatch. Declared type has size = %zu but actual member has size = %zu", sizeof(arg1), sizeof(parent_struct_pointer->arg2[0]));\
            free(parent_struct_pointer);\
            return NULL;\
        }
```
This doesn't leak because we are freeing the pointer if validation fails.

> It seems like all checks should be done before we malloc.

This check validates size discrepancy. `parent_struct_pointer->arg2[0]` is only accessible after malloc. So, we cannot change the order.

Reply:
I have removed this arg validation check now because while reproing this scenario in UTs I realized that complier takes care of arg validation and we do not need to do this explicitly.